draft-ietf-tls-esni.md

```diff
@@ -446,33 +482,112 @@ to harmonize these to make sure that we maintain key separation.]]
+
+- If the server resumed a session or negotiated parameters not based on
+  certificates, the client MUST abort the connection with an illegal_parameter
+  alert. This case is invalid because {{client-behavior}} requires the client
+  to only offer ESNI-established sessions, and {{server-behavior}} requires
+  the server to decline ESNI-established sessions if it did not accept ESNI.
+
+- The client MUST verify that the certificate is valid for ESNIKeys.public_name.
+  If invalid, it MUST abort the connection with the appropriate alert.
+
+- If the server requests a client certificate, the client MUST respond with an
+  empty Certificate message, denoting no client certificate.
+
+Note that verifying a connection for the public name does not verify it for the
+origin. The TLS implementation MUST NOT surface such connections as successful to
+the application. It additionally MUST ignore all session tickets and session IDs
+presented by the server. These connections are only used to trigger retries, as
+described in {{client-behavior}}. This may be implemented, for instance, by
+reporting a failed connection with a dedicated error code.
+
+## Client-Facing Server Behavior {#server-behavior}
 
-If the server does not negotiate TLS 1.3 or above, then the client
-MUST abort the connection with an "unsupported_version" alert.
-If the server does not provide an "encrypted_server_name" extension
-in EncryptedExtensions, the client MUST abort the connection with
-an "illegal_parameter" alert. Moreover, it MUST check that
-ClientESNIInner.nonce matches the value of the
-"encrypted_server_name" extension provided by the server,
-and otherwise abort the connection with an "illegal_parameter"
-alert.
+Upon receiving an "encrypted_server_name" extension, the client-facing
+server MUST check that it is able to negotiate TLS 1.3 or greater. If not,
+it MUST abort the connection with a "handshake_failure" alert.
 
-## Client-Facing Server Behavior
+If the ClientEncryptedSNI.record_digest value does not match the
+cryptographic hash of any known ESNIKeys structure, it MUST ignore the
+extension and proceed with the connection, with the following
+behavior:
 
-Upon receiving an "encrypted_server_name" extension, the client-facing
-server MUST first perform the following checks:
+- It MUST include the "encrypted_server_name" extension in
+  EncryptedExtensions message with the "response_type" field set to
+  "esni_retry_requested" and the "retry_keys" field set to one or more
+  ESNIKeys structures with up-to-date keys. Servers MAY supply multiple
+  ESNIKeys values of different versions. This allows a server to support
+  multiple versions at once.
 
-- If it is unable to negotiate TLS 1.3 or greater, it MUST
-  abort the connection with a "handshake_failure" alert.
+- The server MUST NOT resume any sessions offered by the client that
+  were established with ESNI. Such sessions are associated with the ESNI name,
+  which the server was unable to decrypt.
```

Comment:
I had real trouble with this text because it doesn't say which direction the resumption applies to.  In general, when talking about resumption, you need to be very clear about whether you are talking about accepting a PSK for *this* session, or whether you are talking about maybe later establishing a resumed connection.  In this case, it's possible that both apply, so I had trouble with this.

Reply:
My take is that the text refers to the former, not the latter, though it is a little unclear.

Reply:
It's just the former, yeah.

The intent of the text is to prevent resumption/SNI mismatches. With ESNI and, in particular, this ESNI to SNI fallback, a client offering {(non-GREASE) ESNI, SNI, PSK} now needs to consider all four possible responses:

1. Server ACK'd ESNI and did full handshake
2. Server ACK'd ESNI and resumed PSK
3. Server ignored ESNI (thus used SNI) and did full handshake
4. Server ignored ESNI (thus used SNI) and resumed PSK

In particular, whether the server ACKs or ignores ESNI determines which name the server believes it is acting on behalf of. So we want that to be consistent with the identity the client has associated with the PSK, if any. (1) and (3) are vacuously self-consistent, so that's easy.

As for (2) and (4), presumably the client would not be offering a session good for the public name because it's actually trying to connect to the ESNI name. Thus we want to reject (4) and allow (2). (If we allow (4), then the verify-public-name section needs to decide whether this weird scenario counts as authenticating the replacement ESNI keys and that's weird.) That means the server needs to know not the produce (4).

Now, we could that, if a server ignores the ESNI extension due to key mismatch, it should also ignore the PSK extension. That works, except:
* We want to be robust to server rollbacks and partial deployments. Such a server won't know to ignore the PSK extension.
* Being able to send GREASE extensions is nice and a GREASE client wants to accept (4) since all the names are actually identical all around.

Our thinking was to instead take leaf out of EMS's (and new TLS versions in general) book and partitions the connection/resumption space by whether ESNI was used. We remember a bit in the session state for whether ESNI was used when establishing it. We then say that a client offering ESNI [for real, not GREASE] can only offer ESNI-ful sessions, and a server can only accept ESNI-ful sessions in ESNI-ful handshakes.

This gives all the properties we want:
* Clients offering ESNI will only offer an ESNI-ful session, which is invalid for the server to accept if it declined ESNI. Thus it can reject (4) above.
* This can work with ESNI-oblivious servers on rollback. The server makes the ESNI bit in the session state unparseable by older versions of itself. This is the same sort of thing the server would already have had to do for EMS, or TLS 1.3.
* Clients offering GREASE ESNI will almost always offer ESNI-less sessions (last time I talked to the server, it presumably didn't have ESNI either). The server will decline ESNI (since it's random noise), which means it's allowed to resume ESNI-less sessions.

There isn't an immediate need to constrain ESNI-less sessions, so I didn't in this version of the text. Though maybe a full partition is cleaner? It also occurs to me I have no idea what to do about external PSKs. We could say that if you really want ESNI + external PSK (strange considering that ESNI is usually about picking a certificate anyway...), you have to associate an ESNI bit with the EPSK?? EPSKs are a really underspecified use case that I don't know how to evaluate it.

Anyway, I've tweaked the text slightly and left the above questions as-is for now. Is the new text clearer? / Thoughts on the above?

Reply:
I think the proposed text clears things up. @martinthomson @ekr I'll take this PR to the list to see if there are any objections.